tests/scenario/test_ingress_per_app.py

```diff
@@ -0,0 +1,173 @@
+# GIVEN a charm with ingress impl'd
+# WHEN a relation with traefik is formed
+# THEN traefik's config file's `server` section has all the units listed
+# AND WHEN the charm rescales
+# THEN the traefik config file is updated
+
+import tempfile
+from pathlib import Path
+
+import pytest
+import yaml
+from ops import pebble
+from scenario import Container, Context, Model, Mount, Relation, State
+
+
+@pytest.fixture
+def context(traefik_charm):
+    return Context(charm_type=traefik_charm)
+
+
+@pytest.fixture
+def model():
+    return Model(name="test-model")
+
+
+@pytest.fixture
+def temp_opt():
+    with tempfile.TemporaryDirectory() as td:
+        yield Path(td)
+
+
+@pytest.fixture
+def traefik_container(temp_opt):
+    layer = pebble.Layer(
+        {
+            "summary": "Traefik layer",
+            "description": "Pebble config layer for Traefik",
+            "services": {
+                "traefik": {
+                    "override": "replace",
+                    "summary": "Traefik",
+                    "command": '/bin/sh -c "/usr/bin/traefik | tee /var/log/traefik.log"',
+                    "startup": "enabled",
+                },
+            },
+        }
+    )
+
+    opt = Mount("/opt/", temp_opt)
+
+    return Container(
+        name="traefik",
+        can_connect=True,
+        layers={"traefik": layer},
+        service_status={"traefik": pebble.ServiceStatus.ACTIVE},
+        mounts={"opt": opt},
+    )
+
+
+@pytest.mark.parametrize("port, host", ((80, "1.1.1.1"), (81, "10.1.10.1")))
+@pytest.mark.parametrize("event_name", ("joined", "changed"))
```

Comment:
This would result in a matrix test with combos we're not entirely interested in: do we really spec the same outcome for both `joined` and `changed` given the same context?

Reply:
is there such a thing as too much testing? 

Anywhoos, the ingress libs will emit 'data_provided' on changed, joined and created.
So actually we're missing 'created'.